### ISSUE TITLE
--wrap ignores FULLWIDTH (i.e., CJK!)

### ISSUE THREAD

**Issue by jidanni** [htacg/tidy-html5#835]
The awful truth about
```
       --wrap Integer (68 if unset)
              This option specifies the right margin Tidy uses for line wrapping.

              Tidy tries to wrap lines so that they do not exceed this length.

              Set wrap to 0 (zero) if you want to disable line wrapping.
```
is that it mindlessly assumes each character is one screen column, and
ignores e.g.,

```
$ unicode Ｘ
U+FF38 FULLWIDTH LATIN CAPITAL LETTER X

$ perl -wle '$m = join " ", ( "Ｘ" x 5 ) x 15;
$n = join " ", ( "Y" x 10 ) x 15;
print "<P>$_</P>" for $m, $n;
' | tidy
..."HTML Tidy for HTML5 for Linux version 5.6.0">...
<p>ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ
ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ ＸＸＸＸＸ</p>
<p>YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY
YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY
YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY YYYYYYYYYY</p>
```
i.e., with no consideration of CJK users!

It's as if "those Asian people like using wider characters because in
Asia they probably all use wider monitors anyway."

See perl's Text::CharWidth - Get number of occupied columns of a string on terminal.

**Comment by geoffmcl** [comment]
@jidanni you are right, `libTidy` **mindlessly assumes each character is one screen column**, when applying the `wrap` option...

The current code -

```C
static void AddC( TidyPrintImpl* pprint, uint c, uint string_index)
{
    if ( string_index + 1 >= pprint->lbufsize )
        expand( pprint, string_index + 1 );
    pprint->linebuf[string_index] = c;
}

static uint AddChar( TidyPrintImpl* pprint, uint c )
{
    AddC( pprint, c, pprint->linelen );
    return ++pprint->linelen;
}
```

As can be seen, each character added simply bumps the `linelen` by **one**, and that is true whether the character is U+FF38, or U+0020, or whatever...

Tidy has no mechanism to determine fullwidth characters (like Hiragana or Han Ideograph), which occupy two columns per character, **nor** combining characters (apperaring in ISO-8859-11 Thai, Unicode, and so on), which occupy zero columns per character...

That is, it has no perl `use Text::CharWidth qw(mbwidth mbswidth mblen);` to help it out... which for some reason failed to install in my Windows 10 x64 machine, to try it out - compile failed... but anyway, tidy would need, its own code, or a library, or something...

So what is the `use` base? Could be massive...

But more importantly, who is going to offer code, maybe in a fork, I would help where I can... it might be possible... to rewrite the `pretty print` wrapping code... it takes some understanding of `libTidy`...

As usual, there could certainly be an enhancement of the current docs  [wrap](http://api.html-tidy.org/tidy/quickref_next.html#wrap) option... maybe something like -

```
Wrapping is based on simple chactacter count, each occupying one screen column. 
It does not take into account Asian `fullwidth` characters, like CJK, and maybe other cases... 
There is an issue #835 exploring this. If you can help with the code, jump in, please...
```

Have also added a test case [in_835.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_835.html)... but your perl worked file...

I do not think this is anything to do with monitor width...

Look forward to further feedback, code, PR, etc... thanks...

**Comment by jidanni** [comment]
> but your perl worked file...

You mean "your perl worked fine"?
